### PR TITLE
Add support for Subscription.config.affinity

### PR DESCRIPTION
**Pull request by perdasilva**
Signed-off-by: perdasilva <perdasilva@redhat.com>

**Description of the change:**
Adds possibility to override operator deployment pod affinity.
Pulls in latest commit of api as pseudo-version. Will update to api version once this is good. Subscription.config now contains affinity. To override values:

```yaml
affinity:
  nodeAffinity: // override with new nodeAffinity definition in deployment
    ... 
  podAffinity: {} // remove podAffinity definition from deployment
```

or

```yaml
affinity: {}  // completely remove affinity definition from deployment
```

I used reflections to automatically capture any affinity changes in the API coming from upstream kube

**Motivation for the change:**
Help schedule operators in multiarch environments

**Architectural changes:**
N/A

**Testing remarks:**
Not adding e2e since that would really just test the kube scheduler. Opted for more aggressive unit testing to ensure the spec is modified properly. I've also done some manual testing against a kind cluster with two nodes, manipulating namespace labels and including dummy pods to test out the node, pod, and pod anti- affinity settings. Seems to work.

**Reviewer Checklist**
- [ ] Implementation matches the proposed design, or proposal is updated to match implementation
- [ ] Sufficient unit test coverage
- [ ] Sufficient end-to-end test coverage
- [ ] Bug fixes are accompanied by regression test(s)
- [ ] e2e tests and flake fixes are accompanied evidence of flake testing, e.g. executing the test 100(0) times
- [ ] tech debt/todo is accompanied by issue link(s) in comments in the surrounding code
- [ ] Tests are comprehensible, e.g. Ginkgo DSL is being used appropriately
- [ ] Docs updated or added to `/doc`
- [ ] Commit messages sensible and descriptive
- [ ] Tests marked as `[FLAKE]` are truly flaky and have an issue
- [ ] Code is properly formatted

### CODE REVIEW

**Review by timflannagan — COMMENTED** [review]
Can we split the o-f/api bump from the code changes here?

**Comment by perdasilva** [comment]
> Can we split the o-f/api bump from the code changes here?

For no one but you, @timflannagan

File: pkg/controller/operators/olm/overrides/inject/inject.go [hunk 1]
```diff
@@ -230,3 +230,79 @@ func InjectNodeSelectorIntoDeployment(podSpec *corev1.PodSpec, nodeSelector map[
...
+	// special case: if overriding affinity is empty/default,
+	// nil out the deployment affinity
+	if reflect.DeepEqual(&corev1.Affinity{}, affinity) {
+		podSpec.Affinity = nil
+		return nil
+	}
+
+	// override non nil overriding fields
+	newAffinityValue := reflect.ValueOf(affinity)
+	curAffinityValue := reflect.ValueOf(podSpec.Affinity)
```
Comment (exdx) [inline pkg/controller/operators/olm/overrides/inject/inject.go:270]: Do we need to use `ValueOf` if these types are not interfaces? Couldn't we just dereference them normally?

Comment (perdasilva) [reply pkg/controller/operators/olm/overrides/inject/inject.go]: The point of this code was to handle API changes dynamically. I've updated to remove most of the reflection stuff and make it more readable/maintainable. I've added a unit test to detect whether there are been API changes and fail CI with a helpful message of what to do.

File: pkg/controller/operators/olm/overrides/inject/inject.go [hunk 2]
```diff
@@ -230,3 +230,79 @@ func InjectNodeSelectorIntoDeployment(podSpec *corev1.PodSpec, nodeSelector map[
...
+	if podSpec.Affinity == nil || reflect.DeepEqual(&corev1.Affinity{}, podSpec.Affinity) {
+		podSpec.Affinity = new(corev1.Affinity)
+		*podSpec.Affinity = *affinity
+		return nil
+	}
+
+	// special case: if overriding affinity is empty/default,
+	// nil out the deployment affinity
+	if reflect.DeepEqual(&corev1.Affinity{}, affinity) {
+		podSpec.Affinity = nil
```
[the remark this answers is not in the document]
Comment (perdasilva) [reply pkg/controller/operators/olm/overrides/inject/inject.go:265]: That's basically the semantic of the empty/default override. I.e.

Subscription looking like
```yaml
config:
  affinity: {}
```

should end up as a pod spec w/o affinity. I guess the alternative would be to have something like:

```yaml
affinity: {}
```

in the podSpec. But that felt kinda dangly to me. wdyt?

[the remark this answers is not in the document]
Comment (perdasilva) [reply pkg/controller/operators/olm/overrides/inject/inject.go:265]: I also think this *should* be safe. As `affinity: {}` -> corev1.Affinity{} - right?

**Pull request merged**